src/group_theory/submonoid.lean

```diff
@@ -513,140 +547,73 @@ instance has_one : has_one S := ⟨⟨_, S.one_mem⟩⟩
-| 0 := S.one_mem
-| (n + 1) := S.mul_mem h pow_mem
-
-lemma powers_subset {a : M} (h : a ∈ S) : powers a ≤ S :=
-assume x ⟨n, hx⟩, hx ▸ S.pow_mem h
-
-@[simp] lemma coe_pow (a : S) (n : ℕ) : ((a ^ n : S) : M) = a ^ n :=
-by induction n; simp [*, pow_succ]
-
-end submonoid
-
-namespace add_submonoid
-
-variables (S : add_submonoid A)
-
-/-- The multiples `0, x, 2x, ...` of an element `x` of an `add_monoid M` are an `add_submonoid`. -/
-def multiples (x : A) : add_submonoid A :=
-{ carrier := {y | ∃ n:ℕ, add_monoid.smul n x = y},
-  zero_mem' := ⟨0, add_monoid.zero_smul x⟩,
-  add_mem' := by rintros x₁ x₂ ⟨n₁, rfl⟩ ⟨n₂, rfl⟩; exact ⟨n₁ + n₂, add_monoid.add_smul _ _ _ ⟩ }
-
-/-- An element `x` of an `add_monoid` is in the `add_submonoid` generated by `x`. -/
-lemma multiples.self_mem {x : A} : x ∈ multiples x := ⟨1, add_monoid.one_smul x⟩
-
-lemma smul_mem {a : A} (h : a ∈ S) {n : ℕ} : add_monoid.smul n a ∈ S :=
-submonoid.pow_mem (add_submonoid.to_submonoid S) h
+open lattice
 
-lemma multiples_subset {a : A} (h : a ∈ S) : multiples a ≤ S :=
-submonoid.powers_subset (add_submonoid.to_submonoid S) h
+@[to_additive]
+instance : has_le (submonoid M) := ⟨λ S T, ∀ ⦃x⦄, x ∈ S → x ∈ T⟩
 
-@[simp] lemma coe_smul (a : S) (n : ℕ) : ((add_monoid.smul n a : S) : A) = add_monoid.smul n a :=
-submonoid.coe_pow (add_submonoid.to_submonoid S) a n
+@[to_additive]
+instance : has_lt (submonoid M) := ⟨λ S T, S ≤ T ∧ ¬(T ≤ S)⟩
 
-end add_submonoid
+@[to_additive]
+lemma le_def {S T : submonoid M} : S ≤ T ↔ ∀ ⦃x : M⦄, x ∈ S → x ∈ T := iff.rfl
 
-namespace submonoid
+@[simp, to_additive]
+lemma coe_subset_coe {S T : submonoid M} : (S : set M) ⊆ T ↔ S ≤ T := iff.rfl
 
-variables (S : submonoid M)
+@[simp, to_additive]
+lemma coe_ssubset_coe {S T : submonoid M} : (S : set M) ⊂ T ↔ S < T := iff.rfl
 
 /-- The submonoid `M` of the monoid `M`. -/
 @[to_additive "The `add_submonoid M` of the `add_monoid M`."]
-def univ : submonoid M :=
-{ carrier := set.univ,
-  one_mem' := set.mem_univ 1,
-  mul_mem' := λ _ _ _ _, set.mem_univ _ }
+instance : has_top (submonoid M) :=
+⟨{ carrier := set.univ,
+   one_mem' := set.mem_univ 1,
+   mul_mem' := λ _ _ _ _, set.mem_univ _ }⟩
 
 /-- The trivial submonoid `{1}` of an monoid `M`. -/
 @[to_additive "The trivial `add_submonoid` `{0}` of an `add_monoid` `M`."]
-def bot : submonoid M :=
-{ carrier := {1},
-  one_mem' := set.mem_singleton 1,
-  mul_mem' := λ a b ha hb, by simp * at *}
-
-/-- Submonoids of a monoid are partially ordered (by inclusion). -/
-@[to_additive "The `add_submonoid`s of an `add_monoid` are partially ordered (by inclusion)."]
-instance : partial_order (submonoid M) :=
-partial_order.lift (coe : submonoid M → set M) (λ a b, ext') (by apply_instance)
-
-@[to_additive]
-lemma le_def (p p' : submonoid M) : p ≤ p' ↔ ∀ x ∈ p, x ∈ p' := iff.rfl
-
-open lattice
-
-@[to_additive]
-instance : has_bot (submonoid M) := ⟨submonoid.bot⟩
+instance : has_bot (submonoid M) :=
+⟨{ carrier := {1},
```

Comment:
I think this is a good idea, but I also think it is better to fix it once and for all in core. If we start hacking around this in mathlib, it will be harder to clean up once we fix core.

Reply:
Yes, I'd prefer to fix it in Lean 3.8, not here.